README.md
# logger

## 日志打印器

[API documentation](https://lisfan.github.io/logger/)

## Feature 特性

- 解决提交时因eslint提示console的语句无法通过问题
- 仅在开发环境打印日志，生产环境不打印日志

## Detail 详情

- 在console上包装了一层，支持console的所有的方法（包含部分非标准APi，但不包含未被废弃的API），部分API做了变化和新增加，未提及的保原效果不变，只是在原api上封装了一层进行代理运行，API使用方法可以参考[console API](https://developer.mozilla.org/en-US/docs/Web/API/Console/group)
  - 新增的isActivated、color、enable、disable方法
  - 调整error方法的作用：打印时会抛出错误，阻止脚本执行
  - 调整table方法的作用：如果数据非array或object类型，则使用this.log打印
- 若需要在生产环境下调式日志，可以更改或设置LS离线存储的值
   - localStorage设置`IS_DEV`为true
   - localStorage设置`LOGGER_RULES`配置命名空间规则
- 支持配置整个命名空间是否输出日志
- 支持配置命名空间下某个实例方法是否输出日志

## Install 安装

```bash
npm install -S @~lisfan/logger
```

## Usage 起步

```js
import Logger from '@~lisfan/logger'

// 配置规则
Logger.configRules({
   request:true, // 该命名空间支持打印输出
   request.error:false, // 该命名空间下的error方法不支持打印输出
   response:false // 该命名空间不支持打印输出
})

const logger = new Logger() // 默认打印器，命名空间为`logger`
const loggerRequest = new Logger('request') // 创建命名空间为`request`的打印器
const loggerResponse = new Logger('response')

// 创建打印器，但关闭调试功能
const loggerDebug = new Logger({
   name: 'debug',
   debug: false
})

loggerRequest.log('请求url')    =>    [request]: 请求url
loggerRequest.error('请求url')    =>    // 无内容打印
loggerResponse.error('响应数据')    =>    // 无内容打印
loggerDebug.log('请求url')    =>     // 无内容打印
```

CHANGELOG.md
# CHANGELOG 更新日志

## v1.3.1
### 2017-11-06
- [refactor] 实例的打印开关优先级别调至最低

## v1.3.0-rc
### 2017-11-01
- [fix] 实例方法的组织方式重构，修复方法变成非实例方法的错误


## v1.2.0-rc
### 2017-10-28
- [feature] 增加命名空间规则配置相关的api

## v1.1.0-rc
### 2017-10-17
- [refactor] 小重构
- [feature] 增加实例化时新的配置选项：debug，若设置debug为false，则当前实例不打印日志
- [feature] 增加新API，开启日志输出功能(enable)，和关闭日志输出功能(disable)

### 2017-10-16
- [remove] 移除是否对生产环境的依赖，从离线存储中取值时只判断该值是否已存在

## v1.0.0
## 2017-08-21
- [feature] 从先以**父**命名空间为主判断依据更改为以**子**命名空间为主判断依据，例如关闭全局命名空间`util-http`的日志，但独立开启`utils-http.error`http的错误日志

### 2017-06-19
- [feature] 将打印中涉及dom的打印，都转换成打印其dom属性，而不是dom字符串
- [feature] 增加group()打印方法，按组打印，如果第一个参数是字符串类型，则作为组名称标识，之后的每一个参数都会作为一个单独的打印项，并打印出其数据类型字符串标识

### 2017-06-10
- [feature] 日志打印器的启用方式增加localstoreage热开启
- [feature] 支持子命名空间打印

### 2017-05-18
- [refactor] 改成类定义方式

src/index.js
/**
 * @file 日志打印器
 * @author lisfan <goolisfan@gmail.com>
 * @version 1.3.1
 * @licence MIT
 */

import validation from '@~lisfan/validation'

/**
 * 从`localStorage`的`LOGGER_RULES`键中读取**打印规则**配置，以便可以在生产环境开启日志打印调试
 */
const LOGGER_RULES = JSON.parse(global.localStorage.getItem('LOGGER_RULES')) || {}

/**
 * 从`localStorage`的`IS_DEV`键中读取是否为**开发环境**配置，以便可以在生产环境开启日志打印调试
 */
const IS_DEV = JSON.parse(global.localStorage.getItem('IS_DEV')) || process.env.NODE_ENV === 'development'

// 私有方法
const _actions = {
  /**
   * 打印器工厂函数
   * 查找ls中是否存在打印命名空间配置项，若存在，则进行替换覆盖
   * 判断是否存在子命名空间，依次判断子命名空间的长度
   * @param {Logger} self - Logger实例
   * @param {string} method - 打印方法
   * @param {string} color - 颜色值，web安全色 http://www.w3school.com.cn/tiy/color.asp?color=LightGoldenRodYellow
   * @returns {function} - 返回封装后的的打印方法
   */
  logFactory(self, method, color) {
    return function (...args) {
      return _actions.logProxyRun(self, method, color, ...args)
    }
  },
  /**
   * 代理运行打印方法
   * @param {Logger} self - Logger实例
   * @param {string} method - 打印方法
   * @param {string} color - 打印颜色，颜色值，web安全色 http://www.w3school.com.cn/tiy/color.asp?color=LightGoldenRodYellow
   * @param {...*} args - 其他参数
   * @returns {Logger} - 返回实例自身
   */
  logProxyRun(self, method, color, ...args) {
    if (self.isActivated(method)) {
      let formatStr = `%c[${self.$name}]:%c`

      // 遍历参数列表，找出dom元素，进行转换
      args = args.map((arg) => {
        if (validation.isElement(arg)) {
          return [arg]
        }
        return arg
      })

      /* eslint-disable no-console */
      console[method](formatStr, `color: ${color}`, '', ...args)
      /* eslint-enable no-console */
    }

    return self
  },
  /**
   * 代理运行console方法
   * [注] 内部会进行判断是否允许日志输出
   * @param {Logger} self - Logger实例
   * @param {string} method - 打印方法
   * @param {...*} args - 其他参数
   * @returns {Logger} - 返回实例自身
   */
  proxyRun(self, method, ...args) {
    if (self.isActivated(method)) {
      /* eslint-disable no-console */
      console[method](...args)
      /* eslint-enable no-console */
    }

    return self
  }
}

/**
 * @classdesc
 * 日志打印类
 *
 * @class
 */
class Logger {
  /**
   * 打印器命名空间规则配置项
   * - 可以配置整个命名空间是否输出日志
   * - 也可以配置命名空间下某个实例方法是否输出日志
   *
   * @since 1.2.0
   * @memberOf Logger
   * @static
   * @readonly
   * @property {object} rules - 打印器命名空间规则配置集合
   */
  static rules = LOGGER_RULES

  /**
   * 更改命名空间规则配置项
   * [注]从`localStorage`的`LOGGER_RULES`键中读取规则配置优先级最高，始终会覆盖其他规则
   *
   * @since 1.2.0
   * @static
   * @param {object} rules - 配置参数
   * @param {string} [rules.name] - 日志器命名空间
   * @param {boolean} [rules.debug] - 调试模式是否开启
   * @returns {Logger}
   * @example
   * // 定义规则
   * Logger.configRules = {
   *    utils-http:false // 整个utils-http不可打印输出
   *    utils-calc.log=true // utils-calc打印器的log方法不支持打印输出
   * }
   */
  static configRules(rules) {
    const ctor = this

    ctor.rules = {
      ...ctor.rules,
      ...rules,
      ...LOGGER_RULES,
    }

    return ctor
  }

  /**
   * 默认配置选项
   * 为了在生产环境能开启调试模式
   * 提供了从localStorage获取默认配置项的措施
   *
   * @since 1.0.0
   * @memberOf Logger
   * @readonly
   * @static
   * @property {string} name='logger' - 日志器命名空间，默认为'logger'
   * @property {boolean} debug=true - 调试模式是否开启，默认开启
   */
  static options = {
    name: 'logger',
    debug: true,
  }

  /**
   * 更新默认配置选项
   *
   * @since 1.0.0
   * @static
   * @param {object} options - 配置参数
   * @param {string} [options.name] - 日志器命名空间
   * @param {boolean} [options.debug] - 调试模式是否开启
   * @return {Logger}
   */
  static config(options) {
    const ctor = this

    // 以内置配置为优先
    ctor.options = {
      ...ctor.options,
      ...options
    }

    return ctor
  }

  /**
   * 构造函数
   *
   * @param {object} [options] - 实例配置选项，若参数为`string`类型，则表示设定为`options.name`的值
   * @param {string} [options.name] - 日志器命名空间
   * @param {boolean} [options.debug] - 调试模式是否开启
   */
  constructor(options) {
    const ctor = this.constructor

    if (validation.isString(options)) {
      this.$options = {
        ...ctor.options,
        name: options
      }
    } else {
      this.$options = {
        ...ctor.options,
        ...options
      }
    }
  }

  /**
   * 实例的配置项
   *
   * @since 1.0.0
   * @readonly
   * @returns {object}
   */
  $options = undefined

  /**
   * 获取实例的命名空间配置项
   *
   * @since 1.1.0
   * @readonly
   * @returns {string}
   */
  get $name() {
    return this.$options.name
  }

  /**
   * 获取实例的调试模式配置项
   *
   * @since 1.1.0
   * @returns {string}
   */
  get $debug() {
    return this.$options.debug
  }

  /**
   * 检测当前是否调试模式是否激活：可以打印日志
   *
   * @since 1.1.0
   * @param {string} [method] - 若指定了该参数，则精确检测具体的实例方法
   * @returns {boolean}
   */
  isActivated(method) {
    // 如果不是开发模式
    if (!IS_DEV) {
      return false
    }

    const ctor = this.constructor
    // 以子命名空间的状态优先
    let status = ctor.rules[this.$name]
    // 先判断其子命名空间的状态

    // 如果存在放法名，则判断子命名空间
    // 当前方法名存在子命名空间里且明确设置为false时，则不打印
    // 当前子命名空间如果明确false，则不打印
    if (method) {
      const subStatus = ctor.rules[`${this.$name}.${method}`]

      if (validation.isBoolean(subStatus)) {
        status = subStatus
      }
    }

    // 如果明确指定该命名空间不开启日志打印，则不打印
    if (status === false) {
      return false
    }

    // 最后才判断实例是否禁用了调试
    if (!this.$debug) {
      return false
    }

    return true
  }

  /**
   * 创建一个指定颜色的打印方法
   *
   * @since 1.1.0
   * @param {string} color - 颜色值
   * @returns {Function} - 返回自定义颜色的打印方法
   */
  color(color) {
    return _actions.logFactory(this, 'log', `${color}`)
  }

  /**
   * 启用日志输出
   *
   * @since 1.1.0
   * @returns {Logger}
   */
  enable() {
    this.$options.debug = true
    return this
  }

  /**
   * 禁用日志输出
   *
   * @since 1.1.0
   * @returns {Logger}
   */
  disable() {
    this.$options.debug = false
    return this
  }

  /**
   * 常规日志打印
   *
   * @since 1.0.0
   * @function
   * @param {...*} args - 任意数据
   * @return {Logger}
   */
  log(...args) {
    return _actions.logProxyRun(this, 'log', 'lightseagreen', ...args)
  }

  /**
   * 警告日志打印
   *
   * @since 1.0.0
   * @function
   * @param {...*} args - 任意数据
   * @return {Logger}
   */
  warn(...args) {
    return _actions.logProxyRun(this, 'warn', 'goldenrod', ...args)
  }

  /**
   * 调用栈日志打印
   *
   * @since 1.0.1
   * @function
   * @param {...*} args - 任意数据
   * @return {Logger}
   */
  trace(...args) {
    return _actions.logProxyRun(this, 'trace', 'lightseagreen', ...args)
  }

  /**
   * 错误日志打印，同时会抛出错误，阻塞后续逻辑
   *
   * @since 1.0.0
   * @param {...*} args - 参数列表
   * @throws Error - 抛出错误提示
   */
  error(...args) {
    const message = args.map((value) => {
      return value.toString()
    }).join(' ')

    throw new Error(message)
  }

  /**
   * log的同名方法，使用方法请参考{@link Logger#log}
   *
   * @since 1.1.0
   * @function
   * @param {...*} args - 任意数据
   * @return {Logger}
   * @see Logger#log
   *
   */
  info(...args) {
    return this.log(...args)
  }

  /**
   * log的同名方法，使用方法请参考{@link Logger#log}
   *
   * @since 1.1.0
   * @function
   * @param {...*} args - 任意数据
   * @return {Logger}
   * @see Logger#log
   */
  debug(...args) {
    return this.log(...args)
  }

  /**
   * 区别于console.table
   * - 对象或数组类型数据以表格的方式打印
   * - 若非这两种数据类型，则调用log方法打印
   *
   * @since 1.1.0
   * @param {*} data - 任意数据
   * @returns {Logger}
   */
  table(data) {
    if (validation.isArray(data) && validation.isPlainObject(data)) {
      return this.log(data)
    }

    return _actions.proxyRun(this, 'table', data)
  }

  /**
   * 打印纯对象数据
   *
   * @since 1.1.0
   * @function
   * @param {object} obj - 纯对象数据
   * @return {Logger}
   */
  dir(...args) {
    return _actions.proxyRun(this, 'dir', ...args)
  }

  /**
   * 打印纯对象数据
   *
   * @since 1.1.0
   * @function
   * @param {object} obj - 纯对象数据
   * @return {Logger}
   */
  dirxml(...args) {
    return _actions.proxyRun(this, 'dirxml', ...args)
  }

  /**
   * 创建一个组，接下来所有的打印内容，都会包裹在组内，直到调用groupEnd()方法结束，退出组
   *
   * @since 1.1.0
   * @function
   * @param {string} [label] - 标签名称
   * @return {Logger}
   */
  group(...args) {
    return _actions.proxyRun(this, 'group', ...args)
  }

  /**
   * 类似group()方法，区别在于调用该方法后打印的内容都是折叠的，需要手动展开
   *
   * @since 1.1.0
   * @function
   * @param {string} [label] - 标签名称
   * @return {Logger}
   */
  groupCollapsed(...args) {
    return _actions.proxyRun(this, 'groupCollapsed', ...args)
  }

  /**
   * 关闭组
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  groupEnd(...args) {
    return _actions.proxyRun(this, 'groupEnd', ...args)
  }

  /**
   * 统计被执行的次数
   *
   * @since 1.1.0
   * @function
   * @param {string} [label] - 标签名称
   * @return {Logger}
   */
  count(...args) {
    return _actions.proxyRun(this, 'count', ...args)
  }

  /**
   * 开始设置一个timer追踪操作任意的消耗时间，直到调用timeEnd()结束追踪，消耗时间单位为毫秒
   *
   * @since 1.1.0
   * @function
   * @param {string} label - 标签名称
   * @return {Logger}
   */
  time(...args) {
    return _actions.proxyRun(this, 'time', ...args)
  }

  /**
   * 结束追踪
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  timeEnd(...args) {
    return _actions.proxyRun(this, 'timeEnd', ...args)
  }

  /**
   * 结束追踪
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  timeStamp(...args) {
    return _actions.proxyRun(this, 'timeStamp', ...args)
  }

  /**
   * 开始记录一个性能分析简报，直到调用profileEnd()结束记录
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  profile(...args) {
    return _actions.proxyRun(this, 'profile', ...args)
  }

  /**
   * 结束记录
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  profileEnd(...args) {
    return _actions.proxyRun(this, 'profileEnd', ...args)
  }

  /**
   * 断言表达式，若结果为false，是抛出失败输出
   *
   * @since 1.1.0
   * @function
   * @param {boolean}  assertion - 表达式
   * @param {...*} - 断言失败输出
   * @return {Logger}
   */
  assert(...args) {
    return _actions.proxyRun(this, 'assert', ...args)
  }

  /**
   * 清空控制台
   *
   * @since 1.1.0
   * @function
   * @return {Logger}
   */
  clear(...args) {
    return _actions.proxyRun(this, 'clear', ...args)
  }
}

export default Logger


src/logger.js
/* eslint-disable */

/**
 * 日志打印器 基于大哥 的log改版
 *
 * 模块性质：工具模块
 * 作用范围：pc、mobile
 * 依赖模块：utils/env
 * 来源项目：扫码点单H5
 * 初始发布日期：2017-05-24 20:00
 * 最后更新日期：2017-10-16 13:46
 *
 * ## 特性
 * - [注]只会在开发模式下打印日志内容
 * - 若需要在生产环境下调式日志，可以更改或设置LS离线存储的值：如设置为开发环境，并配置哪些命名空间的日志需要打印
 * - 所以当你在生产环境时，你可以将当前的开发模式开启，然后通过再更改对应的日志打印器命名空间来开启日志打印 * - 可以配置指定的命名空间是否打印内容
 * - 支持配置子命名空间，指定某个方法名才打印
 * - 提供四个方法，打印不同级别的日志内容：log，warn，error，trace
 * - 如果在开发模式则不支持打印，可以通过localstorage更改为开发模式
 *
 * ## Changelog
 * ### 2017-05-18
 * - 改成类定义方式
 *
 * ### 2017-06-10
 * - 日志打印器的启用方式增加localstoreage热开启
 * - 支持子命名空间打印
 *
 * ### 2017-06-19
 * - 将打印中涉及dom的打印，都转换成打印其dom属性，而不是dom字符串
 * - 增加group()打印方法，按组打印，如果第一个参数是字符串类型，则作为组名称标识，之后的每一个参数都会作为一个单独的打印项，并打印出其数据类型字符串标识
 *
 * ### 2017-08-21
 * - 从先以**父**命名空间为主判断依据更改为以**子**命名空间为主判断依据
 * - 例如关闭全局命名空间`util-http`的日志，但独立开启`utils-http.error`http的错误日志
 *
 * ### 2017-10-16
 * - 移除是否对生产环境的依赖，从离线存储中取值时只判断该值是否已存在
 *
 * ## Usage
 * ``` js
 *
 * import Logger from 'utils/logger'
 *
 * Logger.config({
 *    request:true,
 *    request.error:false,
 *    response:false // 该命名空间禁止打印
 * })
 *
 * const loggerRequest = new Logger('request')
 * const loggerResponse = new Logger('response')
 *
 * loggerRequest.log('请求url')    =>    [request]: 请求url
 * loggerRequest.error('请求url')    =>    // 无内容打印
 * loggerResponse.error('响应数据')    =>    // 无内容打印
 * ```
 *
 * @version 1.1.0
 */

import IS_DEV from './env'

// // 变量类型简写对应
// const TYPE_ABBR = {
//   'undefined': 'undef.',
//   'null': 'null',
//   'number': 'num.',
//   'string': 'str.',
//   'boolean': 'bool.',
//   'array': 'arr.',
//   'object': 'obj.',
//   'function': 'fun.',
//   'date': 'date',
//   'regexp': 'reg.',
//   'error': 'err.',
//   'element': 'dom.'
// }

// 私有方法
const _actions = {
  /**
   * 获取变量值的数据类型名称
   * @param {*} value - 变量值
   * @returns {string} 返回数据类型名称
   */
  'typeof': function (value) {
    return Object.prototype.toString.call(value).slice(8, -1)
  },
  /**
   * 获取变量值的数据类型缩写格式
   * @param {*} value - 变量值
   * @returns {string} 返回缩写格式
   */
  // getTypeAbbrStr (value) {
  //   let type = _actions.typeof(value)
  //   // 如果是dom元素节点，只则输出element
  //   if (type.indexOf('element') >= 0) {
  //     type = 'element'
  //   }
  //
  //   return TYPE_ABBR[type]
  // }
}

// 所有日志打印器的命名空间
const LOG_MAP = {
  group: 'color: lightseagreen',
  trace: 'color: lightseagreen',
  log: 'color: lightseagreen',
  warn: 'color: goldenrod',
  error: 'color: red',
}

/**
 * 日志打印类
 * @class Logger 日志打印类
 */
export default class Logger {
  $name = undefined
  $debug = undefined

  /**
   * 构造函数器
   * @constructor
   * @param {object|string} options - 日志器命名空间
   * @param {string} name - 日志器命名空间
   */
  constructor(options) {
    let $options = {
      name: 'logger',
      debug: true
    }

    if (typeof options === 'string') {
      $options.name = options
    } else {
      $options = {
        ...$options,
        ...options
      }
    }

    this.$name = $options.name
    this.$debug = $options.debug
  }

  /**
   * 日志打印命名空间启用输出开关，默认命名空间自动输出日志，设置为false时，可以关才输出
   * @static
   * @enum
   */
  static options = JSON.parse(global.localStorage.getItem('LOGGER')) || {}

  /**
   * 更改全局配置参数
   * @static
   * @param {object} [options={}] - 配置参数
   */
  static config(options = {}) {
    // 以内置配置为优先
    Logger.options = Object.assign(options, Logger.options)
  }
}

/**
 * 在原型上增加实例方法
 */
Object.entries(LOG_MAP).forEach(([method, color]) => {
  // 查找ls中是否存在打印命名空间配置项，若存在，则进行替换覆盖
  // 判断是否存在子命名空间，依次判断子命名空间的长度
  Logger.prototype[method] = function (...args) {
    // 如果实例未开启调试模式
    if (!this.$debug) {
      return false
    }

    // 如果不是开发模式
    if (!IS_DEV) {
      return false
    }

    // 以子命名空间的状态优先
    let status = Logger.options[this.$name]
    // 先判断其子命名空间的状态

    // 当前方法名存在子命名空间里且明确设置为false时，则不打印
    // 当前子命名空间如果明确false，则不打印
    const subStatus = Logger.options[`${this.$name}.${method}`]

    if (typeof subStatus === 'boolean') {
      status = subStatus
    }

    // 如果明确指定该命名空间不开启日志打印，则不打印
    if (status === false) {
      return false
    }

    let formatStr = `%c[${this.$name}]:%c`
    /* eslint-disable no-console */
    if (method === 'group') {
      // 第一个参数为字符串格式时将作为组名
      if (_actions.typeof(args[0]) === 'String') {
        console.group(formatStr, color, '', args[0])
        args.splice(0, 1)
      } else {
        console.group(formatStr, color, '')
      }

      // 换了一种方式，使用group
      // [msl] dir方法需要特殊处理，内部使用console.groupCollapsed实现
      // 参数逐个打印

      const typeStyle = 'color:white;font-size:x-small;background-color:black'
      // 组成新的格式化字符串
      // 和占位参数值数组

      // 数组和对象就不需要打印类型名称了
      args.forEach((arg) => {
        // 提前求得类型缩写
        const typeAbbr = _actions.typeof(arg)
        // 如果参数值是dom节点，则转换成json对象格式输出
        if (_actions.typeof(arg).indexOf('Element') >= 0) {
          arg = [arg]
        }
        console.log('%c%s', typeStyle, typeAbbr, arg)
      })

      console.groupEnd()
    } else {
      // 遍历参数列表，找出dom元素，进行转换
      args = args.map((arg) => {
        if (_actions.typeof(arg).indexOf('Element') >= 0) {
          return [arg]
        }
        return arg
      })

      console[method](formatStr, color, '', ...args)
    }
    /* eslint-enable */
  }
})
